source/MissionPanel.cpp

```diff
@@ -968,25 +968,53 @@ void MissionPanel::Accept(bool force)
 
 	++availableIt;
 	player.AcceptJob(toAccept, GetUI());
-	if(availableIt == available.end() && !available.empty())
+
+	cycleInvolvedIndex = 0;
+	if(available.empty())
+		return;
+	if(availableIt == available.end())
 		--availableIt;
 
 	// Check if any other jobs are available with the same destination. Prefer
 	// jobs that also have the same destination planet.
 	if(toAccept.Destination())
 	{
+		const list<Mission>::const_iterator startHere = availableIt;
 		const Planet *planet = toAccept.Destination();
 		const System *system = planet->GetSystem();
-		for(auto it = available.begin(); it != available.end(); ++it)
+		bool stillLooking = true;
+		for(auto it = startHere; it != available.end(); ++it)
 			if(it->Destination() && it->Destination()->IsInSystem(system))
 			{
-				availableIt = it;
 				if(it->Destination() == planet)
-					break;
+				{
+					availableIt = it;
+					return;
+				}
+				else if(stillLooking)
+				{
+					stillLooking = false;
+					availableIt = it;
+				}
 			}
+		for(auto it = startHere; it != available.begin(); )
+		{
+			--it;
+			if(it->Destination() && it->Destination()->IsInSystem(system))
+			{
+				if(it->Destination() == planet)
+				{
+					availableIt = it;
+					return;
+				}
+				else if(stillLooking)
+				{
+					stillLooking = false;
+					availableIt = it;
+				}
+			}
+		}
```

Comment:
Deduplication
```suggestion
		auto SelectNext = [&stillLooking, &planet, &system, &availableIt](const Mission &mission) -> bool
		{
			if(mission.Destination() && mission.Destination()->IsInSystem(system))
			{
				if(mission.Destination() == planet)
				{
					availableIt = it;
					return true;
				}
				else if(stillLooking)
				{
					stillLooking = false;
					availableIt = it;
				}
			}
			return false;
		};
		for(auto it = startHere; it != available.end(); ++it)
			if(SelectNext(*it))
				return;
		for(auto it = startHere; it != available.begin(); )
		{
			--it;
			if(SelectNext(*it)
				return;
		}
```

Reply:
Neat trick, although I think I would go with slightly different captures (copy is as good as reference for pointers, right?).  Is it bad style to merge the --it into the next statement in this case?